lib/webidl2.js

```diff
@@ -229,35 +231,32 @@
       return name;
     }
 
-    let last_token = null;
+    let consume_position = 0;
 
     function probe(type) {
-      return tokens.length > 0 && tokens[0].type === type;
+      return tokens.length > consume_position && tokens[consume_position].type === type;
     }
 
     function consume(...candidates) {
       // TODO: use const when Servo updates its JS engine
       for (let type of candidates) {
         if (!probe(type)) continue;
-        if (typeof value === "undefined" || tokens[0].value === value) {
-          last_token = tokens.shift();
-          line += count(last_token.trivia, "\n");
-          return last_token;
-        }
+        const token = tokens[consume_position];
+        consume_position++;
+        line += count(token.trivia, "\n");
+        return token;
       }
     }
 
     function unescape(identifier) {
       return identifier.startsWith('_') ? identifier.slice(1) : identifier;
     }
 
-    function unconsume(...toks) {
-      // TODO: use const when Servo updates its JS engine
-      // https://github.com/servo/servo/issues/20231
-      for (let tok of toks) {
-        line -= count(tok.trivia, "\n");
+    function unconsume(position) {
+      while (consume_position > position) {
```

Comment:
Trying to figure out if this can get stuck in a loop and if we should guard against that. I couldn’t come up with anything tho.

Reply:
It should be safe unless `position` is `-Infinity`, which will not happen.

Reply:
yeah, that's what I thought too... but just wanted to make sure :)

Reply:
Thinking twice is great 😁